lib/notificationHandler.js

```diff
@@ -0,0 +1,297 @@
+
+    /**
+     * Add a message to the scope and category
+     *
+     * @param {string} scope - scope of the message
+     * @param {string|null|undefined} category - category of the message, if non we check against regex of scope
+     * @param {string} message - message to add
+     * @param {string|null} instance - instance e.g. hm-rpc.1
+     */
+    addMessage(scope, category, message, instance) {
+        if (!instance) {
+            instance = this.host;
+        } else if (instance !== this.host && !instance.startsWith('system.adapter.')) {
+            instance = 'system.adapter.' + instance;
+        }
+
+        let categories = [category];
+
+        if (!category) {
+            // check regex - in parse text we can have multiple matchings
+            categories = this._parseText(scope, message);
+        }
+
+        for (const _category of categories) {
+            if (_category) {
+                this.currentNotifications[scope] = this.currentNotifications[scope] || {};
+                this.currentNotifications[scope][_category] = this.currentNotifications[scope][_category] || {};
+                // array of all messages for instance/category
+                this.currentNotifications[scope][_category][instance] = this.currentNotifications[scope][_category][instance] || [];
+
+                if (!this.setup[scope] || !this.setup[scope][_category]) {
+                    // no setup for this instance/category combination found - so nothing to add
+                    this.log.warn(`${this.logPrefix} No configuration found for scope "${scope}" and category "${_category}"`);
+                    continue;
+                }
+
+                // if limit exceeded, remove last element - use while if it somehow grew too big
+                while (this.setup[scope][_category].limit < this.currentNotifications[scope][_category][instance].length + 1) {
+                    this.currentNotifications[scope][_category][instance].pop();
+                }
+                // add new element at beginning
+                this.currentNotifications[scope][_category][instance].unshift(message);
+            }
+        }
+    }
+
+    /**
+     * Check the given message against all regular expressions of the given scope
+     *
+     * @param {string} scope - scope of the message
+     * @param {string} message - message to check
+     * @return {any[]}
+     * @private
+     */
+    _parseText(scope, message) {
+        const categories = [];
+        if (this.setup[scope]) {
+            for (const category of Object.keys(this.setup[scope])) {
+                if (!this.setup[scope][category]) {
+                    continue;
+                }
+
+                // check all regular expressions for this category
+                for (const regex of this.setup[scope][category].regex) {
+                    if (regex.test(message)) {
+                        // matching category
+                        categories.push(category);
```

Comment:
and break the loop once one regex matched?

Reply:
yes its only for each category so we can break here